### PR TITLE
(cloud-merge) Support more multi case strategies about cache

### PR DESCRIPTION
**Pull request by Lchangliang**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-common/src/main/java/org/apache/doris/common/Config.java [hunk 1]
```diff
@@ -2685,6 +2685,24 @@ public static boolean isNotCloudMode() {
 
     @ConfField(mutable = true, masterOnly = false)
     public static String security_checker_class_name = "";
+    public static int history_cloud_warm_up_job_keep_max_second = 7 * 24 * 3600;
```
Comment (morrySnow) [inline fe/fe-common/src/main/java/org/apache/doris/common/Config.java:2688]: miss  @ConfField(?
BTW, we should use global session variable as far as possible

File: be/src/cloud/cloud_warm_up_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+void CloudWarmUpManager::handle_jobs() {
+#ifndef BE_TEST
+    while (true) {
+        JobMeta* cur_job = nullptr;
+        {
+            std::unique_lock lock(_mtx);
+            _cond.wait(lock, [this]() { return _closed || !_pending_job_metas.empty(); });
+            if (_closed) break;
```
Comment (deardeng) [inline be/src/cloud/cloud_warm_up_manager.cpp:63]: 这里be close了，还要继续做66行的任务么？看着没有return啊

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudEnv.java [hunk 3]
```diff
@@ -438,4 +452,63 @@ public void replayUpdateCloudReplica(UpdateCloudReplicaInfo info) throws MetaNot
     public boolean getEnableStorageVault() {
         return this.enableStorageVault;
     }
+
+    public long loadCloudWarmUpJob(DataInputStream dis, long checksum) throws Exception {
+        // same as loadAlterJob
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudEnv.java:457]: remove unnecessary comment....

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudEnv.java [hunk 4]
```diff
@@ -438,4 +452,63 @@ public void replayUpdateCloudReplica(UpdateCloudReplicaInfo info) throws MetaNot
...
+
+        Map<Long, CloudWarmUpJob> cloudWarmUpJobs;
+        cloudWarmUpJobs = this.getCacheHotspotMgr().getCloudWarmUpJobs();
+
+        int size = 0;
+        checksum ^= size;
+        dos.writeInt(size);
+
+        checksum ^= size;
+        dos.writeInt(size);
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudEnv.java:500]: why do we write and checksum 0 twice here?
add comment if it is by intent

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudEnv.java [hunk 5]
```diff
@@ -438,4 +452,63 @@ public void replayUpdateCloudReplica(UpdateCloudReplicaInfo info) throws MetaNot
...
+                LOG.info("cloud warm up job is expired, {}, ignore it", cloudWarmUpJob.getJobId());
+                continue;
+            }
+            this.getCacheHotspotMgr().addCloudWarmUpJob(cloudWarmUpJob);
+        }
+        LOG.info("finished replay cloud warm up job from image");
+        return newChecksum;
+    }
+
+    public long saveCloudWarmUpJob(CountingDataOutputStream dos, long checksum) throws IOException {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudEnv.java:489]: make sure the save/load procedures are compactible with the existing warehouse

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudEnv.java [hunk 6]
```diff
@@ -438,4 +452,63 @@ public void replayUpdateCloudReplica(UpdateCloudReplicaInfo info) throws MetaNot
...
+        newChecksum ^= size;
+        for (int i = 0; i < size; i++) {
+            CloudWarmUpJob cloudWarmUpJob = CloudWarmUpJob.read(dis);
+            if (cloudWarmUpJob.isExpire() || cloudWarmUpJob.getJobState() == JobState.DELETED) {
+                LOG.info("cloud warm up job is expired, {}, ignore it", cloudWarmUpJob.getJobId());
+                continue;
+            }
+            this.getCacheHotspotMgr().addCloudWarmUpJob(cloudWarmUpJob);
+        }
+        LOG.info("finished replay cloud warm up job from image");
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudEnv.java:485]: it is not replay, it is just a load from the image.

File: fe/fe-core/src/main/java/org/apache/doris/catalog/OlapTable.java [hunk 7]
```diff
@@ -2024,6 +2024,10 @@ public String getEstimatePartitionSize() {
         return "";
     }
 
+    public boolean containPartition(String partitionName) {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/OlapTable.java:2027]: naming `containPartition` -> `containsPartition`

File: fe/fe-core/src/main/java/org/apache/doris/analysis/WarmUpClusterStmt.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    public WarmUpClusterStmt(String dstClusterName, List<Map<TableName, String>> tableList, boolean isForce) {
+        this.dstClusterName = dstClusterName;
+        this.tableList = tableList;
+        this.isForce = isForce;
+        this.isWarmUpWithTable = true;
+    }
+
+    @Override
+    public void analyze(Analyzer analyzer) throws AnalysisException, UserException {
+        super.analyze(analyzer);
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/WarmUpClusterStmt.java:66]: throw exception if not cloud mode

File: fe/fe-core/src/main/java/org/apache/doris/analysis/ShowCloudWarmUpStmt.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    public long getJobId() {
+        return jobId;
+    }
+
+    public boolean showAllJobs() {
+        return showAllJobs;
+    }
+
+    @Override
+    public void analyze(Analyzer analyzer) throws AnalysisException, UserException {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/ShowCloudWarmUpStmt.java:62]: throw exception if not cloud mode

File: fe/fe-core/src/main/java/org/apache/doris/analysis/ShowCacheHotSpotStmt.java [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    private List<String> whereExprValues = new ArrayList<>();
+    private List<String> whereExpr = new ArrayList<>();
+    private SelectStmt selectStmt;
+
+    public ShowCacheHotSpotStmt(String url) {
+        tablePath = url;
+    }
+
+    @Override
+    public void analyze(Analyzer analyzer) throws UserException {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/ShowCacheHotSpotStmt.java:76]: throw exception if not cloud mode

File: fe/fe-core/src/main/java/org/apache/doris/analysis/CancelCloudWarmUpStmt.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    public CancelCloudWarmUpStmt(Expr whereClause) {
+        this.whereClause = whereClause;
+    }
+
+    public long getJobId() {
+        return jobId;
+    }
+
+    @Override
+    public void analyze(Analyzer analyzer) throws AnalysisException {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/CancelCloudWarmUpStmt.java:36]: throw exception if not cloud mode

File: be/src/cloud/cloud_tablet_hotspot.h
Comment (gavinchou) [inline be/src/cloud/cloud_tablet_hotspot.h:1]: lack of comment

File: be/src/cloud/cloud_tablet_hotspot.cpp
Comment (gavinchou) [inline be/src/cloud/cloud_tablet_hotspot.cpp:1]: lack of comment

File: be/src/cloud/cloud_warm_up_manager.cpp
Comment (gavinchou) [inline be/src/cloud/cloud_warm_up_manager.cpp:1]: lack of comment

File: be/src/cloud/cloud_warm_up_manager.h
Comment (gavinchou) [inline be/src/cloud/cloud_warm_up_manager.h:1]: lack of comment

File: be/src/io/cache/block_file_cache_downloader.cpp
Comment (gavinchou) [inline be/src/io/cache/block_file_cache_downloader.cpp:1]: lack of comment

File: be/src/io/cache/block_file_cache_downloader.h
Comment (gavinchou) [inline be/src/io/cache/block_file_cache_downloader.h:1]: lack of comment

File: be/src/olap/compaction.cpp [hunk 12]
```diff
@@ -1122,6 +1122,8 @@ Status CloudCompactionMixin::construct_output_rowset_writer(RowsetWriterContext&
     ctx.compaction_level =
             _engine.cumu_compaction_policy(compaction_policy)->new_compaction_level(_input_rowsets);
 
+    ctx.write_file_cache = true;
```
Comment (gavinchou) [inline be/src/olap/compaction.cpp:1125]: make it a config option.
base compaction may flush the cache, which cause performance issue

File: be/src/olap/tablet_meta.h [hunk 13]
```diff
@@ -191,6 +191,7 @@ class TabletMeta {
     void revise_delete_bitmap_unlocked(const DeleteBitmap& delete_bitmap);
 
     const std::vector<RowsetMetaSharedPtr>& all_stale_rs_metas() const;
+    std::map<std::string, RowsetMetaSharedPtr> snapshot_rs_metas() const;
```
Comment (gavinchou) [inline be/src/olap/tablet_meta.h:194]: add comment what is returned, what are the key and value represent for

**Review by gavinchou — CHANGES_REQUESTED** [review]

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

File: be/src/cloud/cloud_backend_service.cpp [hunk 14]
```diff
@@ -45,4 +57,137 @@ Status CloudBackendService::create_service(CloudStorageEngine& engine, ExecEnv*
...
+    case TWarmUpTabletsRequestType::SET_JOB: {
+        LOG_INFO("receive the warm up request.")
+                .tag("request_type", "SET_JOB")
+                .tag("job_id", request.job_id);
+        st = manager->check_and_set_job_id(request.job_id);
+        if (!st) {
+            LOG_WARNING("SET_JOB failed.").error(st);
+            break;
+        }
+        [[fallthrough]];
```
[the remark this answers is not in the document]
Comment (Lchangliang) [reply be/src/cloud/cloud_backend_service.cpp:99]: When sending the first RPC, it contains SET_JOB and SET_BATCH params.

File: fe/fe-core/src/main/java/org/apache/doris/common/FeConstants.java [hunk 15]
```diff
@@ -80,6 +80,7 @@ public class FeConstants {
     public static final String FS_PREFIX_FILE = "file";
 
     public static final String INTERNAL_DB_NAME = "__internal_schema";
+    public static final String INTERNAL_FILE_CACHE_HOTSPOT_TABLE_NAME = "cloud_cache_hotspot";
```
[the remark this answers is not in the document]
Comment (Lchangliang) [reply fe/fe-core/src/main/java/org/apache/doris/common/FeConstants.java:83]: The table name is not same as selectdb. But it no problem for compatibility.

File: fe/fe-core/src/main/java/org/apache/doris/cluster/ClusterNamespace.java [hunk 16]
```diff
@@ -32,6 +32,10 @@ public class ClusterNamespace {
 
     private static final String CLUSTER_DELIMITER = ":";
 
+    public static String getFullName(String cluster, String name) {
+        return linkString(cluster, name);
+    }
```
[the remark this answers is not in the document]
Comment (Lchangliang) [reply fe/fe-core/src/main/java/org/apache/doris/cluster/ClusterNamespace.java:37]: Copy from selectdb, remove it.

File: fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudEnv.java (hunk 5 above)
Comment (Lchangliang) [reply fe/fe-core/src/main/java/org/apache/doris/cloud/catalog/CloudEnv.java:489]: The persistence is same.

File: be/src/cloud/config.h [hunk 17]
```diff
@@ -88,4 +88,6 @@ DECLARE_mInt32(schedule_sync_tablets_interval_s);
 // Cloud mow
 DECLARE_mInt32(mow_stream_load_commit_retry_times);
 
+DECLARE_mInt32(sync_load_for_tablets_thread);
```
[the remark this answers is not in the document]
Comment (gavinchou) [reply be/src/cloud/config.h:92]: sync_load_for_tablets_num_threads

**Review by gavinchou — APPROVED** [review]

**Comment by Lchangliang** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/analysis/CancelCloudWarmUpStmt.java (hunk 11 above)
Comment (deardeng) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/CancelCloudWarmUpStmt.java:37]: 这些文件 不要放到这个包目录下？
org/apache/doris/cloud/analysis 可以么？
```suggestion
    public void analyze(Analyzer analyzer) throws AnalysisException {
```

**Comment by Lchangliang** [comment]
run p0

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**